ArtOfIllusion/src/artofillusion/Scene.java

```diff
@@ -132,22 +134,17 @@ public void setTime(double t)
   {
     time = t;
     boolean processed[] = new boolean [objects.size()];
-    for (int i = 0; i < objects.size(); i++)
-      {
-        ObjectInfo info = objects.elementAt(i);
-        applyTracksToObject(info, processed, null, i);
-      }
-    for (ObjectInfo obj : objects)
-      obj.getObject().sceneChanged(obj, this);
+    AtomicInteger index = new AtomicInteger();
```

Comment:
Atomic variable allows me to get and increment counter from effectively final variable inside lambda expression.
I use this ideom there because there was index counter and I leave this logic.
From other side the indexOf uses the same Atomic counter when fill indexes map

Reply:
When you find yourself needing workarounds like that, it's a pretty good indication that the code in question really shouldn't be using a functional style...  When you need access to the loop variable, an explicit loop is usually the clearest way to write it.

Reply:
I'd overall agree, but....

That's the thing... It doesn't *need* the loop variable. The code at line 181 used `indexOf()` to fill the need, and it was perfectly adequate. (Note that the latest commit changed that - I'm not sure why???)

If I understand the comment, "I'm keeping the counter because there was a counter before." Which does not make sense to me.

Reply:
In both  loops we just iterate over all objects list and passing curent object index to applyTrack method. The indexOf returns the same index but do this more expensive. On each call it also checks objectIndexMap on null, create and fill if empty and return obejct index from map or return -1 if not found (which is imposiible in this case)

Reply:
I seriously doubt if the overhead of `indexOf` would be noticeable here, when we consider what else is going on in the method call.

Do you have any test results that suggest it would be significant?

Peter is right, though: One or the other, not this half-and-half thing. (It just looks... wierd)

- If the `indexOf` overhead is an actual bottleneck, stick with a true `for` loop.
- If the overhead is not significant, just stick with the more readable version.

Reply:
Well... I create some performance testcase. All of variants are fast enough, None of them bottlenecks execution. 
The fastest case is to generate indexes with IntegerStream.range() method. (This not mentioned above and in code)
On second place is pure for loop
Atomic counter takes third position.
Getting index from index map is the slowest. I see it 2 to 4 times slower then atomic counter.

Reply:
Were you actually testing versions of *this method* with realistic workloads, or were you doing a micro-benchmark of accessing item indices in this manner?

Either way, the atomic integer option has to go.

Reply:
I create a micro-benchmark.

Reply:
With a proper warm-up? Multiple runs, to average correctly? Testing it with different sizes of load, to see if there are one-time startup costs?

Micro-Benchmarks are really hard to get right.

It would take an absolutely massive number of objects for the overhead to increase to a point where it's significant.

I've tested this in-app:
-------------

- 1e6 cubes (It took some time to even *create* them!) with the default position & rotation tracks, but no keyframes.
- Gave the camera a position & rotation track *with* keyframes, 
- Ran the live animation preview

This is about the ideal test setup, as it forces the app to run `setTime()` many times in succession, and therefore loop through the *million* cubes quite a few times, but does not have the tracks do any complicated math to overshadow effects from the loop itself.

Results:
----------

- *No* difference from how the animation preview usually performs on my system.
- This surprised me, as I was expecting *1 Million cubes* to have at least *some* effect, but the framerate looked about normal, and manually scrubbing the timecourse also responded normally.

Verdict:
-----------

The `indexOf()` version is fine. no need to micro-optimize this.

Reply:
I use java's Micro Benchmark Harness for test so it has proper warm-up and multiple runs to average results, And I use difeerent workloads (from 10 to 10K objects) .
In general all loop variants are fast but other looks simpler for me then indexof.

Reply:
> other looks simpler for me then indexof.

How so? The `IntStream` version looks just as bad to me as the `AtomicInteger`, because they obfuscate the meaning of the field.

The classic for-loop variant at least has the virtue that it is a very commonly recognized idiom, due to many years as the only way to do this.

Reply:
From my POV the indexOf usage also obfuscates the fact that we just use simple sequential index over our objects...

I do not see the big trouble to understand what IntStream do. I see range from 0 to objects size and I see index variable. Very similar to classic style "for" idiom.
Also I see in IntStream usage benefit in performance...
As I measured in my benchmark it works faster then other

Reply:
>indexOf usage also obfuscates the fact that we just use simple sequential index over our objects...

But that's an artifact of the implementation, not something critical about how this is supposed to work. When reading this code, I don't need to have it shoved in my face that we're using indexes, because it *does not matter*.

>I see in IntStream usage benefit in performance...

What kind of real clock times are you talking about? My own tests showed that the *user* Will. Not. Notice.